### PR TITLE
Binary union implementation

### PR DESCRIPTION
**Pull request by scferrada**

### CODE REVIEW

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBinaryUnion.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import se.liu.ida.hefquin.engine.queryplan.executable.IntermediateResultBlock;
+import se.liu.ida.hefquin.engine.queryplan.executable.IntermediateResultElementSink;
+import se.liu.ida.hefquin.engine.queryproc.ExecutionContext;
+
+public class ExecOpBinaryUnion implements BinaryExecutableOp{
+
+	@Override
+	public int preferredInputBlockSize() {
+		// Since this algorithm processes the input solution mappings
+		// sequentially (one at a time), and input block size of 1 may
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBinaryUnion.java:13]:
```suggestion
		// sequentially (one at a time), an input block size of 1 may

```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBinaryUnion.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	@Override
+	public boolean requiresCompleteChild1InputFirst() {
+		return false;
+	}
+
+	@Override
+	public void processBlockFromChild1(final IntermediateResultBlock input, final IntermediateResultElementSink sink,
+			ExecutionContext execCxt) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBinaryUnion.java:25]:
```suggestion
			final ExecutionContext execCxt) {

```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBinaryUnion.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	@Override
+	public void wrapUpForChild1(IntermediateResultElementSink sink, ExecutionContext execCxt) {
+		// nothing to be done here
+	}
+
+	@Override
+	public void processBlockFromChild2(final IntermediateResultBlock input, final IntermediateResultElementSink sink,
+			ExecutionContext execCxt) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBinaryUnion.java:37]:
```suggestion
			final ExecutionContext execCxt) {

```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import se.liu.ida.hefquin.engine.queryplan.executable.impl.ops.ExecOpBinaryUnion;
+import se.liu.ida.hefquin.engine.queryplan.logical.BinaryLogicalOp;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpUnion;
+import se.liu.ida.hefquin.engine.queryplan.physical.BinaryPhysicalOpForLogicalOp;
+
+public class PhysicalOpBinaryUnion implements BinaryPhysicalOpForLogicalOp{
+	
+	protected final LogicalOpUnion lop;
+
+	protected PhysicalOpBinaryUnion(LogicalOpUnion lop) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java:19]:
```suggestion
	protected PhysicalOpBinaryUnion( final LogicalOpUnion lop ) {

```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+		assert lop != null;
+		this.lop = lop;
+	}
+
+	@Override
+	public ExpectedVariables getExpectedVariables(ExpectedVariables... inputVars) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java:25]:
```suggestion
	public ExpectedVariables getExpectedVariables( final ExpectedVariables... inputVars ) {

```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+		final Set<Var> certainVars = new HashSet<>( inputVars[0].getCertainVariables());
+		final Set<Var> possibleVars = new HashSet<>( inputVars[0].getPossibleVariables() );
+
+		certainVars.addAll(inputVars[1].getCertainVariables());
+		possibleVars.addAll(inputVars[1].getPossibleVariables());
+		possibleVars.removeAll(certainVars);
+
+		return new ExpectedVariables() {
+			@Override public Set<Var> getCertainVariables() { return certainVars;}
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java:38]: That's incorrect. The set of certain variables of a union operator is the _intersection_ of the sets of certain variables of the two inputs (rather than the union of these two sets, as in your current implementation).

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+			@Override public Set<Var> getPossibleVariables() { return possibleVars;}
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java:39]: That's incorrect as well. The set of possible variables in this case is the union of _all_ variables from both inputs (i.e., certain and possible ones) minus the variables that are certain variables of the union operator.

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		};
+	}
+
+	@Override
+	public BinaryExecutableOp createExecOp(ExpectedVariables... inputVars) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java:44]:
```suggestion
	public BinaryExecutableOp createExecOp( final ExpectedVariables... inputVars ) {

```

**Review by hartig — CHANGES_REQUESTED** [review]

**Comment by hartig** [comment]
I have reviewed the PR. See my comments above. Notice that several of these comments are a "Suggested change" which you can commit directly in the Web interface of github. Hence, you don't need to go to your IDE and reimplement these suggestions manually. It is a neat feature of github ;-)     ...however, after you have accepted these changes in the Web interface, you may have to do a `git pull` on your computer in order to get these changes into your branch.

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java (hunk 6 above)
Comment (scferrada) [reply src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java:40]: So, my question now is regarding the semantics of the union operator. At first I thought that it would be similar to SPARQL union, where you can have a query like this, taken from the SPARQL specification:
```
SELECT ?x ?y
WHERE  { { ?book dc10:title ?x } UNION { ?book dc11:title  ?y } }
```
Where the certain variables are the union of the input variables. So, please, can you show me what I'm getting wrong?

Comment (hartig) [reply src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java:38]: Yes, our union operator works exactly like union in SPARQL. I assume you have read the definition of "certain variables" and "possible variables" that I had added in the comments in the ExpectedVariables interface. Certain variables of an operator are all the variables that can be expected to be bound in _every_ solution mapping produced by that operator. In your example, that would be only ?book but not ?x and not ?y. In general, it is the intersection of the certain variables of the two input operators.

Comment (scferrada) [reply src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java:38]: Right, I get it now, thanks!

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+					assertEquals(2, b.size());
+					assertNotEquals(null, b.get(y));
+					assertEquals("http://example.org/y3", b.get(y).getURI());
+					b3Found = true;
+				}
+			}
+			else if (b.get(y) != null && b.get(y).getURI().equals("http://example.org/y1")) {
+				assertEquals(1, b.size());
+				b4Found = true;
+			}
```
Comment (hartig) [inline src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java:142]: I think there should be an `else` branch here in order to capture faulty behavior.

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+				b1Found = true;
+			}
+			else if (b.get(x).getURI().equals("http://example.org/x3")) {
+				assertEquals("http://example.org/y3", b.get(y).getURI());
+				b2Found = true;
+			} 
+			else if (b.get(x).getURI().equals("http://example.org/x2")) {
+				assertEquals("http://example.org/y2", b.get(y).getURI());
+				b3Found++;
+			}
```
Comment (hartig) [inline src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java:204]: Also here, there should be an `else` branch with a `fail(...)`.

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+				b4Found = true;
+			}
+		}
+		assertTrue(b1Found);
+		assertTrue(b2Found);
+		assertTrue(b3Found);
+		assertTrue(b4Found);
+	}
+	
+	protected void _testKeepUnionDuplicates() {
```
Comment (hartig) [inline src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java:150]: Can you please add another such unit test where the duplicates come from the same input. Just to make sure that implementations get this option right as well.

**Review by hartig — CHANGES_REQUESTED** [review]

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java (hunk 6 above)
Comment (hartig) [reply src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBinaryUnion.java:38]: This looks good now. Thanks for fixing!

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+	
+		assertFalse( it.hasNext() );
+	
+		boolean b1Found = false;
+		boolean b2Found = false;
+		boolean b3Found = false;
+		boolean b4Found = false;
+		
+		for ( final Binding b : result ) {
+			if (b.get(x) != null) {
```
Comment (hartig) [inline src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java:121]:
```suggestion
			if ( b.contains(x) ) {
```

Comment (hartig) [reply src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java:121]: Sorry, I didn't see that before.

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+					assertEquals(2, b.size());
+					assertNotEquals(null, b.get(y));
+					assertEquals("http://example.org/y3", b.get(y).getURI());
+					b3Found = true;
+				}
+				else {
+					fail("Unexpected URI for ?x: " + b.get(x).getURI());
+				}
+			}
+			else if (b.get(y) != null && b.get(y).getURI().equals("http://example.org/y1")) {
```
Comment (hartig) [inline src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java:142]:
```suggestion
			else if ( b.contains(y) && b.get(y).getURI().equals("http://example.org/y1") ) {
```

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java (hunk 11 above)
Comment (hartig) [reply src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java:156]: Thanks for adding the extra unit test !

**Review by hartig — CHANGES_REQUESTED** [review]

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java (hunk 12 above)
Comment (hartig) [reply src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java]: I am committing this myself now. I don't want to delay you further ;-)

**Comment by hartig** [comment]
Great, now it's ready. I will merge it.

**Review by hartig — APPROVED** [review]
